### PR TITLE
Add experiment slug substitution to the feature config

### PR DESCRIPTION
**Pull request by jhugman**
Fixes [EXP-3659](https://mozilla-hub.atlassian.net/browse/EXP-3659).

This implements [Option 3 in ADR0012](https://github.com/mozilla/experimenter/blob/jhugman/exp-3630-adr-for-exposure-events-of-coenrolling-features/docs/adrs/0012-exposure-events-and-coenrolling-features.md).

More concretely:

- the `replace_str` function recursively crawls through the FeatureConfig at enrollment, changing any string with `{experiment}` to the experiment slug.
- adds `recordExperimentExposure(slug: String)` to `FeatureHolder` in both Android and iOS.

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/nimbus/src/json.rs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+///
+/// This recursively descends into the object, looking at string values and keys.
+pub(crate) fn replace_str_in_map(map: &mut Map<String, Value>, from: &str, to: &str) {
+    let replacer = |s: &str| -> Option<String> {
+        if s.contains(from) {
+            Some(s.replace(from, to))
+        } else {
+            None
+        }
+    };
```
Comment (jeddai) [inline components/nimbus/src/json.rs:33]: It looks like this code is written the same in a couple places — could it be refactored out into the file scope?

File: components/nimbus/src/json.rs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    for k in map.keys() {
+        if let Some(new) = replacer(k) {
+            changes.insert(k.to_owned(), new);
+        }
+    }
+
+    for (k, new) in changes {
+        let v = map.remove(&k).unwrap();
+        _ = map.insert(new, v);
+    }
```
Comment (jeddai) [inline components/nimbus/src/json.rs:82]: Is this effectively supporting the following?
```js
{
    '{experiment}': 'someval'
}
```

File: components/nimbus/src/json.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                "not": true,
+                "or": 2,
+            })
+        );
+    }
+
+    #[test]
+    fn test_replace_str_in_map_keys() {
+        let mut value = json!({
+            "{test}-en-US": "{test}",
```
Comment (jeddai) [inline components/nimbus/src/json.rs:141]: Ah yep, neat!

**Review by jeddai — APPROVED** [review]
r+wc

It looks great!

File: components/nimbus/src/enrollment.rs [hunk 4]
```diff
@@ -1150,12 +1151,16 @@ fn get_enrolled_feature_configs(
...
-    let branch_features = match &experiment.get_branch(branch_slug) {
+    let mut branch_features = match &experiment.get_branch(branch_slug) {
         Some(branch) => branch.get_feature_configs(),
         _ => Default::default(),
     };
 
-    let branch_feature_ids = branch_features
+    branch_features.iter_mut().for_each(|f| {
+        json::replace_str_in_map(&mut f.value, SLUG_REPLACEMENT_PATTERN, experiment_slug);
+    });
```
[the remark this answers is not in the document]
Comment (jhugman) [reply components/nimbus/src/enrollment.rs:1161]: I'm not sure I understand.

[the remark this answers is not in the document]
Comment (jeddai) [reply components/nimbus/src/enrollment.rs:1161]: Just commenting that I get a feeling this particular use case of replacing strings in experiment values is going to grow more than where it's currently at

**Pull request merged**